test/extended/dns/dns.go

```diff
@@ -381,4 +460,92 @@ var _ = Describe("[sig-network-edge] DNS", func() {
 		pod := createDNSPod(f.Namespace.Name, cmd)
 		validateDNSResults(f, pod, expect, times)
 	})
+
+	It("should answer A and AAAA queries for a dual-stack service", func() {
+		// Only run this test on dual-stack enabled clusters.
+		networkConfig, err := oc.AdminConfigClient().ConfigV1().Networks().Get(context.Background(), "cluster", metav1.GetOptions{})
+		if err != nil {
+			e2e.Failf("unable to get cluster network config: %v", err)
+		}
+		usingIPv4 := false
+		usingIPv6 := false
+		for _, clusterNetworkEntry := range networkConfig.Status.ClusterNetwork {
+			addr, _, err := net.ParseCIDR(clusterNetworkEntry.CIDR)
+			if err != nil {
+				continue
+			}
+			if addr.To4() != nil {
+				usingIPv4 = true
+			} else {
+				usingIPv6 = true
+			}
+		}
+
+		if !usingIPv4 || !usingIPv6 {
+			Skip("skipping test on non dual-stack enabled platform")
+		}
+
+		By("creating a dual-stack service on a dual-stack cluster")
+
+		ctx := context.Background()
+		serviceName := "v4v6"
+		createOpts := metav1.CreateOptions{}
+		service, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(ctx, createDualStackServiceSpec(serviceName, false, "", nil), createOpts)
+		if err != nil {
+			e2e.Failf("unable to create dual-stack service: %v", err)
+		}
+
+		v4 := createEndpointSliceSpec("dns-test", serviceName, discoveryv1beta1.AddressTypeIPv4)
+		if _, err := f.ClientSet.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).Create(ctx, v4, createOpts); err != nil {
+			e2e.Failf("unable to create endpointslice %s: %v", v4.Name, err)
+		}
+
+		v6 := createEndpointSliceSpec("dns-test", serviceName, discoveryv1beta1.AddressTypeIPv6)
+		if _, err := f.ClientSet.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).Create(ctx, v6, createOpts); err != nil {
+			e2e.Failf("unable to create endpointslice %s: %v", v6.Name, err)
+		}
+
+		v4ips := ipsForEndpointSlice(v4)
+		v6ips := ipsForEndpointSlice(v6)
+
+		var v4ClusterIP string
+		var v6ClusterIP string
+
+		for i, ipFamily := range service.Spec.IPFamilies {
+			if ipFamily == kapiv1.IPv4Protocol {
+				v4ClusterIP = service.Spec.ClusterIPs[i]
+			} else if ipFamily == kapiv1.IPv6Protocol {
+				v6ClusterIP = service.Spec.ClusterIPs[i]
```

Comment:
The Service API godoc says that the entries in `spec.ipFamilies` and `spec.clusterIPs` "must correspond" and that each field has a maximum of 2 entries; is it safe to assume that means the ordering is the same?

Reply:
> is it safe to assume that means the ordering is the same?

That's how I interpreted the documentation.

Reply:
Assumptions make me nervous, but worst case, we get spurious CI failures and need to revise the logic, so
/lgtm